src/EditorFeatures/Core/Diagnostics/DiagnosticsClassificationTaggerProvider.cs

```diff
@@ -62,18 +61,61 @@ protected internal override bool IsEnabled
             => !_editorOptionsService.Factory.GlobalOptions.GetOptionValue(DefaultTextViewHostOptions.IsInContrastModeId);
 
         protected internal override bool IncludeDiagnostic(DiagnosticData data)
-            => data.CustomTags.Contains(WellKnownDiagnosticTags.Unnecessary);
+        {
+            // Fading diagnostic might have an 'Unnecessary' custom tag or
+            // unnecessary additional location indices.
+            if (!data.CustomTags.Contains(WellKnownDiagnosticTags.Unnecessary)
+                && !TryGetUnnecessaryLocationIndices(data, out _))
+            {
+                return false;
+            }
+
+            // Do not fade if user has disabled the fading option corresponding to this diagnostic.
+            PerLanguageOption2<bool> fadingOption;
+            switch (data.Id)
+            {
+                case IDEDiagnosticIds.RemoveUnnecessaryImportsDiagnosticId:
+                    fadingOption = IdeAnalyzerOptionsStorage.FadeOutUnusedImports;
+                    break;
```

Comment:
We could potentially define an extensibility point and MEF import diagnostic-specific classification customization.

Reply:
> These options are now effectively options of the tagger (they only control presentation), not options of the analyzer

This is correct. I'll move them out of IdeAnalyzerOptionsStorage.

@CyrusNajmabadi - The analyzers reporting fading diagnostics should _always_ add the Unnecessary custom tag OR the Unnecessary property in the properties bag with encoded unnecessary additional locations to tag. I think it will be extremely uncommon that we want to add new user visible options to allow users to turn off fading. Unnecessary usings and unreachable code are the only ones that provide user visible options to turn off fading, and personally I don't know if these are useful either.

Reply:
> @CyrusNajmabadi - The analyzers reporting fading diagnostics should always add the Unnecessary custom tag OR the Unnecessary property in the properties bag with encoded unnecessary additional locations to tag. 

My point is that these new features should at least have to think about it.  Having a contract call here force that.

>  I think it will be extremely uncommon that we want to add new user visible options to allow users to turn off fading.

Reply:
I discussed this offline with @CyrusNajmabadi and we decided to introduce a new type `AbstractBuiltInUnnecessaryCodeStyleAnalyzer` which is for all analyzers that report unnecessary code diagnostics. The constructor for this type requires a fading option (or null) and we generate a mapping from diagnostic ID to fading option in `IDEDiagnosticIdToOptionMappingHelper`. The tagger no longer needs to hard code diagnostic IDs or fading options, it invokes `IDEDiagnosticIdToOptionMappingHelper.TryGetMappedFadingOption(diagnosticId)` to check if fading is disabled for the diagnostic.

Reply:
I'm not sure this is a better approach. The problem now is that the fading options are defined in code style layer, but they are global options that only affect client presentation. We also define the mapping of diagnostic id to the option in an analyzer. I think it should be the client defining the mapping from diagnostic id to the option they want to control the presentation of the diagnostic.

Let's consider another example. Let's say a client wants to use a specific squiggle color for certain kind of compiler errors. It shouldn't be the compiler that defines that mapping.

Reply:
Let's discuss offline and try to decide the best approach here.